src/api/baseapi.cpp

```diff
@@ -146,6 +146,10 @@ static void ExtractFontName(const char* filename, std::string* fontname) {
  */
 static void addAvailableLanguages(const std::string &datadir,
                                   std::vector<std::string> *langs) {
+  if (!std::filesystem::exists(datadir)) {
```

Comment:
This check is not sufficient. It passes if `datadir` exists and is a file (which still raises an exception). I suggest to catch the exception which is thrown by `std::filesystem::recursive_directory_iterator`.

Reply:
https://github.com/zdenop/tesseract/blob/2e5a114bd2aab932b53e103d1391e7f27d2d8cff/src/ccutil/ccutil.cpp#L95-L97

This causes that `datadir` is always handled as directory.

Reply:
On macOS:
```
% touch a.txt
% tesseract --list-langs --tessdata-dir a.txt                       
libc++abi: terminating due to uncaught exception of type std::__1::__fs::filesystem::filesystem_error: filesystem error: in recursive_directory_iterator: Not a directory ["a.txt/"]
zsh: abort      tesseract --list-langs --tessdata-dir a.txt
```

Reply:
We don't have main wrapped into try catch?
Error text is ok, we need only catch std::exception globally.

Reply:
@stweil : I tested this PR on linux  (g++ 11.3.0) and it does not produce "uncaught exception" (without egorpugin commit). Do you still need  to add something like this?
```diff 
diff --git a/src/api/baseapi.cpp b/src/api/baseapi.cpp
index 667ead67..e684c7fe 100644
--- a/src/api/baseapi.cpp
+++ b/src/api/baseapi.cpp
@@ -146,6 +146,10 @@ static void ExtractFontName(const char* filename, std::string* fontname) {
  */
 static void addAvailableLanguages(const std::string &datadir,
                                   std::vector<std::string> *langs) {
+  if (!std::filesystem::is_directory(datadir)) {
+    std::cerr << "Error: '" << datadir << "' is not directory.\n";
+    return;
+  }
   if (!std::filesystem::exists(datadir)) {
     std::cerr << "Error: The directory '" << datadir << "' does not exist.\n";
     return;
```

Or should be make these checks at [ccutils.cpp](https://github.com/zdenop/tesseract/blob/2e5a114bd2aab932b53e103d1391e7f27d2d8cff/src/ccutil/ccutil.cpp#L44) and fail if directory not exists?

Reply:
This changes the error messages.

Old
```
libc++abi: terminating due to uncaught exception of type std::__1::__fs::filesystem::filesystem_error: filesystem error: in recursive_directory_iterator: No such file or directory ["tessdata/"]
libc++abi: terminating due to uncaught exception of type std::__1::__fs::filesystem::filesystem_error: filesystem error: in recursive_directory_iterator: Not a directory ["tessdata/"]
```

New:
```
exception: filesystem error: in recursive_directory_iterator: Not a directory ["tessdata/"]
exception: filesystem error: in recursive_directory_iterator: No such file or directory ["tessdata/"]
```

The new error messages are nicer, but are they user friendly? And do we want to introduce C++ exceptions in the C / C++ API?

Reply:
1. C API should be removed later.
2. We want c++ exceptions in api. Nothing wrong here.